### ISSUE TITLE
main/v5.0: mpirun issues finding prterun

### ISSUE THREAD

**Issue by bwbarrett** [open-mpi/ompi#10252]
This is a spin-off from https://github.com/open-mpi/ompi/pull/10224.  There are a couple of issues with mpirun finding the underlying prterun today.  This issue tracks the required fixes:

1) First, there's a typo in ompi_setup_prrte.m4 which causes us to set the search path wrong
2) Second, when that goes wrong, the error message around prterun not found is wrong.  YOu'll see something like:

```
The mpirun ("/opt/amazon/openmpi/bin/prterun") cmd failed to exec its actual executable - your application will NOT execute. Error: No such file or directory
```

But the path is actually the path to mpirun, not the path to where we think prterun should be.

**Comment by shijin-aws** [comment]
One interesting thing I observed is I get different results when using absolute path or relative path to kick off `mpirun`:

```
[ec2-user@ip-172-31-62-242 ompi]$ ./install/bin/mpirun --version
mpirun (Open MPI) 5.1.0a1

Report bugs to https://www.open-mpi.org/community/help/
[ec2-user@ip-172-31-62-242 ompi]$ /home/ec2-user/ompi/install/bin/mpirun --version
The mpirun ("/home/ec2-user/ompi/install/bin/prterun") cmd failed to exec its actual executable - your application will NOT execute. Error: No such file or directory
```

**Comment by bwbarrett** [comment]
yeah, i understand what's happening.  It's broken; going to take some time to figure out how to fix it.

**Comment by shijin-aws** [comment]
I think the issue is here: https://github.com/open-mpi/ompi/blob/main/ompi/tools/mpirun/main.c#L88-L101

If prefix is given, the code use the same ompi prefix to find prterun, but for external prte build it shouldn't? I added this patch 

```
     } else {
         truepath = opal_os_path(0, pfx, "prterun", NULL);
+#if !OMPI_USING_INTERNAL_PRRTE
+        // if OMPI_PRTERUN_PATH is available, try that
+        // for external builds
+        if(0 != strlen(OMPI_PRTERUN_PATH)) {
+            truepath = opal_os_path(0, OMPI_PRTERUN_PATH, "prterun", NULL);
+        }
+#endif
         free(pfx);
     }
```
Then the error is mitigated.

**Comment by bwbarrett** [comment]
Your patch probably works, but just papers over the core problem, which is that the logic about picking the path is wrong.  And even if it was right, it highlights that everything else about setting the prefixes and behaviors is wrong.  I'm working on a proposal for what the code should do, because we're so far off from logical behavior that it's not worth looking at code until we agree on a desired behavior.

**Comment by bwbarrett** [comment]
The core problem with the mpirun wrapper is that we don't seem to have a clear plan around handling prefix behavior.  There are two cases in play: pmix/prrte built as part of OMPI and pmix/prrte built outside OMPI.

In the built-in case, the behavior of the prefix behavior is relatively straight forward, in that the paths are all the same.  So if mpirun had a `--prefix` argument, was called with with a full path, or `OPAL_PREFIX` was set, we can use those to configure the installdirs infrastructure in OMPI and then use that to generate the bindir for prterun.

In the non-builtin case, the behavior of prefix is not straight-forward.  I'm not sure we have the right interface into `prterun` today, but I think what we want is to use `--prefix`, a full path, or `OPAL_PREFIX` to generate a libdir that we pass to prterun to add to LD_LIBRARY_PATH on the remote side (which will mirror the OMPI part of that behavior from 4.1 and previous).  But that still leaves us with what to do about PRRTE.  We don't know where it is (nor do we know where PMIX is).  My proposal is:

  1) Add a --prte-prefix argument to mpirun, which is translated to --prefix when calling prterun.
  2) Add a --pmix-prefix argument to mpirun, which is translated to a libdir to add to prtetun.
  3) If OMPI_PRTERUN is set,  use that as the underlying prterun
  3) Else, run a search based on the configure --with-prte path, the environment, and prte_prefix/bin (we can't use installdirs, because our installdirs configuration might be different than PRRTEs (ie, bindir might be ${execprefix}/bin for ompi and ${prefix}/weird-bin-dir for prrte).

Just some thoughts...

**Comment by bwbarrett** [comment]
Mpirun behavior

In main today, mpirun is a wrapper around prterun.  In an ideal situation, mpirun could be as simple as:

```
setenv("PRTE_MCA_schizo_proxy", "ompi", 1);
setenv("OMPI_VERSION", OMPI_VERSION, 1);
char *base_tool_name = opal_basename(argv[0]);
setenv("OMPI_TOOL_NAME", base_tool_name, 1);
free(base_tool_name);

path_to_prterun = opal_path_findv("prterun", X_OK, environ, NULL);
execve(path_to_prterun, argv, environ);
```

There is some added complexity today in searching for prterun and handling –prefix.  The current cod mixes OMPI’s prefix with PRRTE’s prefix in a way that doesn’t generally work.  How relocation and –prefix work are somewhat dependent on whether or not PMIX and PRRTE are built as part of Open MPI.  Let’s look at both cases independently.
### Built-in PMIx and PRRTE
This is the straight-forward case (for mpirun).  All three packages will necessarily share the same $bindir and $libdir.  So we can be sloppy with who sets which prefix, because they’re all the same.  In this case, the behavior should be:

- If OPAL_PREFIX is set, set PRRTE_PREFIX and PMIX_PREFIX to match.
- Initialize the installdirs code and generate $bindir.
- Set path_to_prterun (above) to $bindir/prterun.

In an ideal world, we could leave argv[0] as the original argv[0] and let prterun sort it out (if we were smart, we could probably get rid of the OMPI_TOOL_NAME thing at the same time).  However, barring that, we need to be a little intelligent with argv[0] and basically set it to s/mpirun/prterun/, leaving the full path so that prterun has the same prefix behavior as mpirun used to w.r.t full path calling.
### External PMIx and PRRTE
This is the problematic case.  PRRTE has a –prefix option, just like mpirun used to.  However, the meaning of –prefix to PRRTE is where to find the PRRTE executables (and libraries).  Setting prrte’s –prefix to OMPI’s install dir will not be a path to enlightenment.  PRRTE has this problem with locating its PMIx as well, although I believe it relies on runpath to find PMIx.
I think the right solution is to add two arguments to PRRTE: –prefix-extra-path and –prefix-extra-ld_library_path which, as the name implies, add things to PATH and LD_LIBRARY_PATH at the launch daemon level.  Mpirun detects a prefix situation by parsing its args, and sets –prefix-extra-path and –prefix-extra-ld_library_path when it calls prterun.  It eats any arguments to –prefix (or sets them to the value of –prte_prefix).  We can add a –prte-prefix option if we’re feeling extra spicy which is translated to –prefix when calling prterun.
Barring a value of –prte_prefix, the next question is how to find prterun.  My suggestion is the following logic:

- If OMPI_PRTERUN is set, use that (abort if not found)
- If PRTE_PREFIX is set, look in ${PRTE_PREFIX}/bin/prterun (abort if not found)
- Add a configure argument –with-prrte-bindir, defaulting to ${with_prrte}/bin.  If ${with_prrte_bindir}/prterun doesn’t exist at configure time and the user specified a path in ${with_prrte}, bail configure.  We now have a PRRTE_BINDIR path we can optionally define.  If defined, look for prterun in PRRTE_BINDIR (abort if not found)
- Search in PATH (abort if not found)

Note that we have no way to set PMIX_PREFIX and PRTE_PREFIX directly; the user must do that themselves.
There’s a question if we need most of the –prefix behavior.  OMPI no longer has backend libraries, so it does not need to modify the backend path.  PMIX does not have backend libraries, so it does not need to modify the backend path.  Both do have relocatable libraries, however, and –prefix can be used to essentially force an update (or a relocation from where application expects them based on a runpath), which is useful for ISV applications.  So I lean towards needing it.

**Comment by bwbarrett** [comment]
I posted a PR for most of the changes in https://github.com/open-mpi/ompi/pull/10321.  This patch doesn't do anything with prefix option, other than pass it on to the underlying PRRTE.  This is probably not what we want, but since I'm not sure what we want, it's the simplest thing to do.  I think we should merge the PR and sort out prefix later.

**Comment by jsquyres** [comment]
Looks like the fix has been merged into both main and v5.0.x -- ok to close this?

**Comment by awlauria** [comment]
I believe it can be closed. If that's not correct please reopen.
